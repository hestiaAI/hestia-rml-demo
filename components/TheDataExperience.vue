<template>
  <div>
    <VRow>
      <VCol cols="12 mx-auto" sm="6">
        <UnitIntroduction
          v-bind="{ companyName: title, dataPortal, isGenericViewer }"
        />
      </VCol>
    </VRow>
    <VRow>
      <VCol cols="12 mx-auto" sm="6">
        <UnitFiles
          v-bind="{
            extensions,
            files,
            multiple,
            allowMissingFiles,
            samples: data,
            isGenericViewer
          }"
          @update="onUnitFilesUpdate"
        />
        <template v-if="progress">
          <BaseProgressCircular class="mr-2" />
          <span>Processing files...</span>
        </template>
        <template v-else-if="error || success">
          <VAlert
            :type="error ? 'error' : 'success'"
            border="top"
            colored-border
            max-width="600"
            >{{ message }}
          </VAlert>
        </template>
      </VCol>
    </VRow>
    <template v-if="success">
      <VRow>
        <VCol>
          <VTabs
            v-model="tab"
            dark
            background-color="primary"
            slider-color="secondary"
            slider-size="4"
            show-arrows
            center-active
            centered
            fixed-tabs
          >
            <VTab>Files</VTab>
            <VTab v-for="(el, index) in defaultView" :key="index">
              {{ el.title }}
            </VTab>
            <VTab v-if="consentForm">Share my data</VTab>
          </VTabs>
          <VTabsItems v-model="tab">
            <VTabItem>
              <UnitFileExplorer v-bind="{ fileManager }" />
            </VTabItem>
            <VTabItem
              v-for="(defaultViewElements, index) in defaultView"
              :key="index"
            >
              <UnitQuery
                v-bind="{
                  defaultViewElements,
                  customPipeline:
                    customPipelines !== undefined
                      ? customPipelines[defaultViewElements.customPipeline]
                      : undefined,
                  sparqlQuery: queries[index],
                  sql: sqlQueries[index],
                  fileManager,
                  postprocessors,
                  index,
                  vega
                }"
                @update="onQueryUpdate"
              />
            </VTabItem>
            <VTabItem v-if="consentForm">
              <UnitConsentForm
                v-bind="{
                  consentForm,
                  allResults,
                  defaultView,
                  fileManager,
                  showDataExplorer
                }"
              />
            </VTabItem>
          </VTabsItems>
        </VCol>
      </VRow>
    </template>
  </div>
</template>

<script>
import { validExtensions } from '~/manifests/utils'
import FileManager from '~/utils/file-manager'
import fileManagerWorkers from '~/utils/file-manager-workers'
import parseYarrrml from '~/utils/parse-yarrrml'
import rdfUtils from '~/utils/rdf'

export default {
  name: 'TheDataExperience',
  props: {
    title: {
      type: String,
      required: true
    },
    dataPortal: {
      type: String,
      default: ''
    },
    data: {
      type: Array,
      default: () => []
    },
    ext: {
      type: String,
      required: true,
      validator(val) {
        return (
          val === 'all' ||
          val.split(',').every(v => validExtensions.includes(v))
        )
      }
    },
    files: {
      type: Array,
      default: () => []
    },
    defaultView: {
      type: Array,
      default: () => []
    },
    multiple: {
      type: Boolean,
      default: false
    },
    preprocessors: {
      type: Object,
      default: () => {}
    },
    postprocessors: {
      type: Object,
      default: undefined
    },
    allowMissingFiles: {
      type: Boolean,
      default: false
    },
    customPipelines: {
      type: Object,
      default: undefined
    },
    isGenericViewer: {
      type: Boolean,
      default: false
    },
    showDataExplorer: {
      type: Boolean,
      default: true
    },
    sparql: {
      type: Object,
      default: () => {}
    },
    sql: {
      type: Object,
      default: () => {}
    },
    vega: {
      type: Object,
      default: () => {}
    },
    yarrrml: {
      type: String,
      default: ''
    },
    databaseBuilder: {
      type: Function,
      default: undefined
    }
  },
  data() {
    return {
      tab: null,
      progress: false,
      error: false,
      success: false,
      message: '',
      rml: '',
      allResults: [...Array(this.defaultView.length)],
      allFiles: null,
      fileManager: new FileManager(
        this.preprocessors,
        this.allowMissingFiles,
        fileManagerWorkers
      ),
      db: null
    }
  },
  computed: {
    queries() {
      return this.defaultView.map(o => this.sparql[o.query])
    },
    sqlQueries() {
      return this.defaultView.map(o => this.sql[o.sql])
    },
    isRdfNeeded() {
      return this.defaultView.filter(v => 'query' in v).length > 0
    },
    consentForm() {
      const consent = this.$store.state.config.consent
      if (consent) {
        const key = this.$route.params.key
        if (key in consent) {
          return consent[key]
        } else if ('default' in consent) {
          return consent.default
        }
      }
      return null
    },
    extensions() {
      return this.ext === 'all'
        ? []
        : this.ext
            .replace(/\s/g, '')
            .split(',')
            .map(ext => `.${ext}`)
    }
  },
  methods: {
    handleError(error) {
      console.error(error)
      this.error = true
      this.message = error instanceof Error ? error.message : error
      this.progress = false
    },
    async onUnitFilesUpdate(uppyFiles) {
      this.message = ''
      this.error = false
      this.success = false
      this.progress = true
      const start = new Date()

      await this.fileManager.init(uppyFiles, this.multiple)
      let processedFiles
      try {
        processedFiles = await this.fileManager.preprocessFiles(this.files)
      } catch (error) {
        this.handleError(error)
        return
      }

      // Populate database
      if (this.databaseBuilder !== undefined) {
        await this.databaseBuilder(this.fileManager)
      }

      if (this.isRdfNeeded && this.yarrrml) {
        try {
          this.rml = await parseYarrrml(this.yarrrml)
          await rdfUtils.generateRDF(this.rml, processedFiles)
        } catch (error) {
          this.handleError(error)
          return
        }
      }

      this.progress = false
      this.success = true

      const elapsed = new Date() - start
      this.message = `Successfully processed in ${elapsed / 1000} sec.`
    },
    onQueryUpdate({ index, result }) {
      // we need to change the object or vue won't see the change
      const newResults = this.allResults.slice()
      newResults[index] = JSON.stringify(result)
      this.allResults = newResults
    }
  }
}
</script>
