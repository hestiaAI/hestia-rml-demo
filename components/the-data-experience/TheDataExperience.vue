<template>
  <client-only placeholder="Loading...">
    <keep-alive>
      <TheDataExperienceDefault v-bind="props">
        <template #unit-files="{ update, ...slotProps }">
          <UnitFiles
            v-bind="{
              extensions,
              files,
              multiple,
              allowMissingFiles,
              samples: data,
              isGenericViewer,
              ...slotProps
            }"
            @update="update"
          >
          </UnitFiles>
        </template>
      </TheDataExperienceDefault>
    </keep-alive>
  </client-only>
</template>

<script>
import { validExtensions } from '@/manifests/utils'

export default {
  name: 'TheDataExperience',
  props: {
    title: {
      type: String,
      required: true
    },
    dataPortal: {
      type: String,
      default: ''
    },
    data: {
      type: Array,
      default: () => []
    },
    ext: {
      type: String,
      required: true,
      validator(val) {
        return (
          val === 'all' ||
          val.split(',').every(v => validExtensions.includes(v))
        )
      }
    },
    files: {
      type: Array,
      default: () => []
    },
    defaultView: {
      type: Array,
      default: () => []
    },
    multiple: {
      type: Boolean,
      default: false
    },
    preprocessors: {
      type: Object,
      default: () => {}
    },
    postprocessors: {
      type: Object,
      default: undefined
    },
    allowMissingFiles: {
      type: Boolean,
      default: false
    },
    customPipelines: {
      type: Object,
      default: undefined
    },
    isGenericViewer: {
      type: Boolean,
      default: false
    },
    showDataExplorer: {
      type: Boolean,
      default: true
    },
    sparql: {
      type: Object,
      default: () => {}
    },
    sql: {
      type: Object,
      default: () => {}
    },
    vega: {
      type: Object,
      default: () => {}
    },
    yarrrml: {
      type: String,
      default: ''
    },
    databaseBuilder: {
      type: Function,
      default: undefined
    }
  },
  computed: {
    extensions() {
      return this.ext === 'all'
        ? []
        : this.ext
            .replace(/\s/g, '')
            .split(',')
            .map(ext => `.${ext}`)
    },
    props() {
      const propNames = [
        'title',
        'dataPortal',
        'defaultView',
        'customPipelines',
        'preprocessors',
        'postprocessors',
        'isGenericViewer',
        'showDataExplorer',
        'files',
        'multiple',
        'allowMissingFiles',
        'sparql',
        'sql',
        'vega',
        'yarrrml',
        'databaseBuilder'
      ]
      return Object.fromEntries(propNames.map(k => [k, this[k]]))
    }
  }
}
</script>
