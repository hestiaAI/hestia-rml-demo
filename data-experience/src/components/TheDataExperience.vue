<template>
  <div :lang="$i18n.locale">
    <SettingsSpeedDial />
    <VMenu v-if="!siteConfigMerged.i18nLocale && locales.length > 1" offset-y absolute>
      <template #activator="{ on, attrs }">
        <BaseButton
          v-bind="attrs"
          mdi-icon="mdiTranslate"
          color="secondary"
          :outlined="false"
          fab
          fixed
          bottom
          left
          v-on="on"
        />
      </template>
      <VList>
        <VListItemGroup v-model="localeIndex" @change="setLocale">
          <VListItem
            v-for="({ code, name }) in locales"
            :key="code"
          >
            <VListItemTitle>{{ name }}</VListItemTitle>
          </VListItem>
        </VListItemGroup>
      </VList>
    </VMenu>
    <template v-if="showLogin">
      <UnitLogin @success="loginSuccessful = true" />
    </template>
    <template v-else>
      <VRow>
        <VItemGroup
          v-model="window"
          mandatory
          class="shrink mx-6 pt-9"
        >
          <VItem
            v-for="i in windows"
            :key="i.id"
            v-slot="{ active, toggle }"
          >
            <div>
              <BaseButtonWindow
                :mdi-icon="i.mdiIcon"
                :tooltip="$t(`${i.id}.name`)"
                :input-value="active"
                @click="toggle"
              />
            </div>
          </VItem>
        </VItemGroup>
        <VCol>
          <VWindow
            v-model="window"
            vertical
          >
            <BaseWindowItem>
              <UnitDownload
                v-if="bubbleConfig && bubbleConfig.dataFromBubble"
                v-bind="{
                  progress,
                  error,
                  success,
                  message
                }"
                @update="onUnitFilesUpdate"
              />
              <UnitIntroduction
                v-else
                v-bind="{
                  progress,
                  error,
                  success,
                  message
                }"
                ref="unit-introduction"
                @update="onUnitFilesUpdate"
              />
            </BaseWindowItem>
            <BaseWindowItem v-if="!experienceConfig.hideFileExplorer">
              <UnitSummary v-if="!experienceConfig.hideSummary" />
              <UnitFileExplorer />
            </BaseWindowItem>
            <BaseWindowItem>
              <VTabs
      v-model="tab"
        dark
        background-color="primary"
        slider-color="secondary"
        slider-size="4"
        show-arrows
        center-active
        centered
        fixed-tabs
        class="fixed-tabs-bar"
        @change="scrollToTop"
      >
        <VTab
          v-for="({ title, disabled }, index) in tabs"
          :key="index"
          :disabled="disabled"
        >
          {{ title }}
        </VTab>
      </VTabs>
      <VTabsItems v-model="tab">
        <div
          v-for="{ id, viewBlock } in tabs"
          :key="`${experienceConfig.slug}-${id}`"
          :data-id="id"
        >
          <VTabItem
            :transition="false"
          >
            <VCol cols="12 mx-auto">
              <VOverlay :value="overlay" absolute opacity="0.8">
                <div
                class="d-flex flex-column align-center"
                style="width: 100%; height: 100%"
                >
                <div class="mb-3">
                    {{ $t('This might take a moment') }}
                  </div>
                  <BaseProgressCircular size="64" width="4" />
                </div>
              </VOverlay>
              <UnitQuery v-bind="viewBlock" />
            </VCol>
          </VTabItem>
        </div>
      </VTabsItems>
            </BaseWindowItem>
            <BaseWindowItem v-if="consent">
              <UnitConsentForm />
            </BaseWindowItem>
          </VWindow>
        </VCol>
      </VRow>
      <!-- <VStepper v-model="step" vertical>

          <VStepperStep complete editable step="1" edit-icon="$vuetify.icons.mdiFileUpload">
            <VIcon>$vuetify.icons.mdiFileUpload</VIcon>
            {{ $t('Load your data') }}
          </VStepperStep>
      </VStepper> -->

    </template>
  </div>
</template>

<script>
import { debounce, cloneDeep, merge } from 'lodash-es'
import { json, timeFormatDefaultLocale } from 'd3'

import { locales, localeCodes } from '@/i18n/locales'

import { mapState, mapMutations } from '@/utils/store-helper'
import DBMS from '@/utils/sql'
import FileManager from '@/utils/file-manager'
import fileManagerWorkers from '@/utils/file-manager-workers'

import BaseButton from '@/components/base/button/BaseButton.vue'
import BaseButtonWindow from '@/components/base/button/BaseButtonWindow.vue'
import BaseProgressCircular from '@/components/base/BaseProgressCircular.vue'
import BaseWindowItem from '@/components/base/BaseWindowItem.vue'
import UnitFileExplorer from '@/components/unit/file-explorer/UnitFileExplorer.vue'
import UnitLogin from '@/components/unit/UnitLogin.vue'
import UnitDownload from '@/components/unit/UnitDownload.vue'
import UnitIntroduction from '@/components/unit/UnitIntroduction.vue'
import UnitQuery from '@/components/unit/UnitQuery.vue'
import UnitSummary from '@/components/unit/UnitSummary.vue'
import SettingsSpeedDial from '@/components/misc/SettingsSpeedDial.vue'
import UnitConsentForm from '@/components/unit/consent-form/UnitConsentForm.vue'

import vuetifyEn from 'vuetify/lib/locale/en'
import vuetifyFr from 'vuetify/lib/locale/fr'

import defaultEn from '@/locales/en.json'
import defaultFr from '@/locales/fr.json'

const messagesDefault = {
  en: {
    ...defaultEn,
    $vuetify: vuetifyEn
  },
  fr: {
    ...defaultFr,
    $vuetify: vuetifyFr
  }
}
const messagesOverride = {
  en: {},
  fr: {
    $vuetify: {
      dataIterator: {
        noResultsText: 'Aucun rÃ©sultat'
      }
    }
  }
}

const siteConfigDefault = {
  i18nLocales: localeCodes,
  messages: {}
}

async function d3Locale({ iso }) {
  const locale = await json(`https://cdn.jsdelivr.net/npm/d3-time-format@4/locale/${iso}.json`)
  timeFormatDefaultLocale(locale)
}

export default {
  name: 'TheDataExperience',
  components: {
    BaseButton,
    BaseButtonWindow,
    BaseProgressCircular,
    BaseWindowItem,
    SettingsSpeedDial,
    UnitLogin,
    UnitDownload,
    UnitFileExplorer,
    UnitIntroduction,
    UnitQuery,
    UnitSummary,
    UnitConsentForm
  },
  props: {
    experienceConfig: {
      type: Object,
      required: true
    },
    siteConfig: {
      type: Object,
      required: true
    },
    bubbleConfig: {
      type: Object,
      required: false
    }
  },
  data() {
    return {
      localeIndex: 0,
      window: 0,
      tab: 0,
      fab: false,
      progress: false,
      error: false,
      success: false,
      message: '',
      overlay: false,
      locales,
      loginSuccessful: false
    }
  },
  computed: {
    ...mapState(['fileManager', 'bubbleCodeword', 'currentTab']),
    ...mapState({ experienceProgress: 'progress' }),
    siteConfigMerged() {
      return cloneDeep(merge(siteConfigDefault, this.siteConfig))
    },
    showLogin() {
      const config = this.bubbleConfig
      return config && !config.bypassLogin && !this.loginSuccessful
    },
    consent() {
      return this.bubbleConfig?.consent
    },
    windows() {
      const windows = [
        {
          id: 'load-data',
          mdiIcon: 'mdiFileUpload'
        }
      ]
      if (!this.experienceConfig.hideFileExplorer) {
        windows.push({
          id: 'file-explorer',
          mdiIcon: 'mdiFileSearch'
        })
      }
      windows.push(
        {
          id: 'understand-data',
          mdiIcon: 'mdiChartBar'
        }
      )
      if (this.consent) {
        windows.push({
          id: 'share-data',
          mdiIcon: 'mdiShare'
        })
      }
      return windows
    },
    tabs() {
      const { viewBlocks } = this.experienceConfig
      const disabled = !this.success || this.experienceProgress
      const tabs = viewBlocks.map(viewBlock => ({
        title: viewBlock.title,
        id: viewBlock.id,
        titleKey: this.k(viewBlock.id),
        disabled,
        viewBlock
      }))
      return tabs.map(({ title, titleKey, ...rest }) =>
        ({ title: this.$tev(titleKey, title), ...rest })
      )
    }
  },
  watch: {
    experienceConfig: {
      immediate: true,
      handler(value) {
        this.setExperienceConfig(cloneDeep(value))
      }
    },
    bubbleConfig: {
      immediate: true,
      handler(value) {
        this.setBubbleConfig(cloneDeep(value))
      }
    },
    siteConfigMerged: {
      immediate: true,
      async handler(value) {
        if (!value.i18nLocale) {
          await this.setLocale()
        }
        // override light theme colors
        if (value.theme) {
          Object.assign(this.$vuetify.theme.themes.light, value.theme)
        }
        this.setSiteConfig(cloneDeep(value))
      }
    },
    fileManager(value) {
      if (value === null) {
        this.setCurrentTab('load-data')
        this.success = false
        this.progress = false
        this.error = false
      }
    },
    // debounce overlay
    experienceProgress: {
      immediate: true,
      handler: debounce(function(value) {
        this.overlay = value
      }, 200)
    },
    tab(tabIndex) {
      const idx = typeof tabIndex !== 'undefined' ? tabIndex : 0
      const { id } = this.tabs[idx]
      if (id !== this.currentTab) {
        this.setCurrentTab(id)
      }
    },
    currentTab(currentTab) {
      this.tab = this.tabs.findIndex(tab => tab.id === currentTab)
    },
    '$i18n.locale'() {
      this.clearStore()
    }
  },
  created() {
    this.siteConfigMerged.i18nLocales.forEach((locale) => {
      /* default messages */
      this.$i18n.mergeLocaleMessage(locale, messagesDefault[locale])
      /* overriding messages */
      this.$i18n.mergeLocaleMessage(locale, messagesOverride[locale])
    })
    this.$watch(
      vm => [vm.experienceConfig, vm.siteConfig, vm.bubbleConfig],
      async() => {
        // clear login state
        this.loginSuccessful = false
        this.clearStore()
        await this.mergeMessages()
      },
      {
        immediate: true
      }
    )
  },
  methods: {
    ...mapMutations([
      'setSiteConfig',
      'setBubbleConfig',
      'setExperienceConfig',
      'setConsentForm',
      'clearStore',
      'setFileManager',
      'setCurrentDB',
      'setCurrentTab'
    ]),
    async setLocale() {
      // locale was switched by the user (or is being initialized)
      const localeProperties = locales[this.localeIndex]
      // -> update d3 locale
      if (process.env.NODE_ENV !== 'test') {
        await d3Locale(localeProperties)
      }
      // -> update vue-i18n locale
      this.$i18n.locale = localeProperties.code
    },
    async mergeMessages() {
      const { messages: messagesExperience = {}, slug } = this.experienceConfig
      const { messages: messagesCustomConfig = {}, i18nLocales, i18nUrl } = this.siteConfigMerged
      let messagesCustomRemote = {}
      if (i18nUrl) {
        // fetch messages to override default messages
        messagesCustomRemote = await json(i18nUrl)
      }
      i18nLocales.forEach((locale) => {
        /* experience messages */
        const messagesObject = { experiences: { [slug]: messagesExperience[locale] } }
        this.$i18n.mergeLocaleMessage(locale, messagesObject)

        /* custom messages (override anything previously merged) */
        if (locale in messagesCustomConfig) {
          // from config
          this.$i18n.mergeLocaleMessage(locale, messagesCustomConfig[locale])
        }

        if (locale in messagesCustomRemote) {
          // from remote endpoint
          this.$i18n.mergeLocaleMessage(locale, messagesCustomRemote[locale])
        }
      })
    },
    // Convert local translation key to global vue18n
    k(localKey) {
      return `experiences.${this.experienceConfig.slug}.viewBlocks.${localKey}.title`
    },
    scrollToTop() {
      window.setTimeout(() => window.scrollTo(0, 0), 10)
    },
    handleError(error) {
      console.error(error)
      this.error = true
      this.message = error instanceof Error ? error.message : error
      this.progress = false
    },
    async onUnitFilesUpdate({ uppyFiles }) {
      const start = new Date()
      const consoleLabel = (...labels) => `TheDataExperience.onUnitFilesUpdate${labels.length ? '.' + labels.join('.') : ''}`
      console.time(consoleLabel())

      const {
        consent,
        experienceConfig: {
          slug,
          databaseConfig,
          files,
          preprocessors,
          keepOnlyFiles,
          viewBlocks
        }
      } = this
      this.message = ''
      this.error = false
      this.success = false
      this.progress = true

      // Clean vuex state
      this.clearStore()

      // Set consent form
      const cForm = consent?.[slug] || consent?.default
      const consentForm = cForm && JSON.parse(JSON.stringify(cForm))
      if (consentForm) {
        const section = consentForm.find(({ type }) => type === 'data')
        section.titles = viewBlocks.map(e => e.title)
        section.ids = viewBlocks.map(e => e.id)
      }
      this.setConsentForm(consentForm)

      try {
        console.time(consoleLabel('initFileManager'))
        // Set file manager
        const fileManager = new FileManager(
          preprocessors,
          fileManagerWorkers,
          files,
          keepOnlyFiles
        )
        await fileManager.init(uppyFiles)
        this.setFileManager(fileManager)
        console.timeEnd(consoleLabel('initFileManager'))

        if (databaseConfig) {
          const consoleLabelDBMS = consoleLabel.bind(null, 'DBMS')
          console.time(consoleLabelDBMS())
          // create database
          console.time(consoleLabelDBMS('createDB'))
          const db = await DBMS.createDB(databaseConfig)
          console.timeEnd(consoleLabelDBMS('createDB'))
          // generate database records via the file manager
          console.time(consoleLabelDBMS('generateRecords'))
          const records = await DBMS.generateRecords(fileManager, databaseConfig)
          console.timeEnd(consoleLabelDBMS('generateRecords'))
          // insert the records into the database
          console.time(consoleLabelDBMS('insertRecords'))
          DBMS.insertRecords(db, records)
          console.timeEnd(consoleLabelDBMS('insertRecords'))
          // commit the database to the Vuex store
          this.setCurrentDB(db)
          console.timeEnd(consoleLabelDBMS())
        }
      } catch (e) {
        this.handleError(e)
        return
      }

      this.progress = false
      this.success = true

      const elapsed = new Date() - start
      this.message = this.$t('successful-processing', { seconds: elapsed / 1000 })

      window.setTimeout(
        // switch to the second tab
        () => this.setCurrentTab(this.tabs[1].id),
        // leave some time for the user to read the success message
        500
      )
    }
  }
}
</script>

<style scoped>
.v-tabs-items .v-window-item {
  min-height: calc(100vh - 48px);
}
</style>
