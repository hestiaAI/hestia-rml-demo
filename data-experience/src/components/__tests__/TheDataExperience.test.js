/**
 * @jest-environment jsdom
 */
import fs from 'fs'
import path from 'path'
import { mount, createLocalVue } from '@vue/test-utils'
import Vuex from 'vuex'
import TheDataExperience from '~/components/TheDataExperience'
import defaultMessages from '~/locales/en.json'
import xpStore from '~/store'
import resizeObserverPolyfill from 'resize-observer-polyfill'

// Mock the Vuex store
const localVue = createLocalVue()
localVue.use(Vuex)

let store

// We need to disable workers for these tests
jest.mock('~/utils/file-manager-workers', () => { })
const originalScrollTo = window.scrollTo
const originalFetch = window.fetch

const propsData = {
  experienceConfig: {
    dataPortal: '',
    dataPortalHtml: '',
    dataPortalMessage: '',
    dataSamples: [],
    tutorialVideos: [],
    videoHeight: 0,
    viewBlocks: [{}],
    files: [],
    slug: 'a-slug'
  },
  siteConfig: {
    experiences: []
  },
  bubbleConfig: {
    id: 'the-data-experience-test',
    apiUrl: 'http://localhost:0000',
    bypassLogin: true,
    experiences: []
  }
}

const mocks = {
  $vuetify: {
    breakpoint: {}
  },
  $i18n: {
    mergeLocaleMessage() {}
  },
  $t: msg => defaultMessages[msg],
  $te: msg => defaultMessages[msg],
  $tev: msg => defaultMessages[msg],
  $tet: msg => defaultMessages[msg]
}

beforeAll(() => {
  // mock fetch for d3.json()
  window.fetch = jest.fn(() => Promise.resolve({ json: () => 'test' }))
  window.scrollTo = jest.fn()
  // for uppy
  globalThis.ResizeObserver = resizeObserverPolyfill.default || resizeObserverPolyfill
})

afterAll(() => {
  delete globalThis.ResizeObserver
})

beforeEach(() => {
  // eslint-disable-next-line import/no-named-as-default-member
  store = new Vuex.Store({
    modules: {
      xp: xpStore
    }
  })
})

afterAll(() => {
  window.scrollTo = originalScrollTo
  window.fetch = originalFetch
})

test('mounts without error', async() => {
  const wrapper = mount(TheDataExperience, {
    propsData,
    store,
    localVue,
    mocks
  })
  expect(wrapper.exists()).toBeTruthy()
  // For some reason it doesn't find the component if we search by name or by import
  expect(
    wrapper.findComponent('[data-id="window-load-data"]').exists()
  ).toBeTruthy()
})

test('data experience html corresponds to snapshot', () => {
  const wrapper = mount(TheDataExperience, {
    propsData,
    store,
    localVue,
    mocks
  })
  expect(wrapper.html()).toMatchSnapshot()
})

test('process simple text file', async() => {
  const wrapper = mount(TheDataExperience, {
    propsData,
    store,
    localVue,
    mocks
  })
  const fileContent = fs.readFileSync(
    path.resolve(__dirname, 'data/test.txt'),
    'utf8'
  )
  const file = new File([fileContent], 'test.txt')

  // Trigger the data processing function
  const unitIntroduction = wrapper.findComponent({ ref: 'unit-introduction' })
  unitIntroduction.vm.$emit('update', { uppyFiles: [file] })

  // I don't know why we need so many ticks
  await wrapper.vm.$nextTick()
  await wrapper.vm.$nextTick()
  await wrapper.vm.$nextTick()
  expect(wrapper.vm.$data.error).toBeFalsy()
  expect(wrapper.vm.$data.success).toBeTruthy()
})
