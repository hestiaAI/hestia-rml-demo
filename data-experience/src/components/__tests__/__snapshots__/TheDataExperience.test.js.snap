// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`data experience html corresponds to snapshot 1`] = `
"<div lang=\\"en\\">
  <vspeeddial direction=\\"top\\" transition=\\"slide-y-reverse-transition\\" fixed=\\"\\" style=\\"z-index: 100; bottom: 8px; right: 8px;\\">
    <vtooltip left=\\"true\\"> <span>Clear all</span></vtooltip>
    <vtooltip left=\\"true\\" fab=\\"\\" dark=\\"\\" small=\\"\\" color=\\"blue darken-2\\" tooltip=\\"Share\\" vbtn-class=\\"\\"> <span>Share</span></vtooltip>
    <vmenu offset-y=\\"\\" absolute=\\"\\">
      <vlist>
        <vlistitemgroup>
          <vlistitem>
            <vlistitemtitle>English</vlistitemtitle>
          </vlistitem>
          <vlistitem>
            <vlistitemtitle>Fran√ßais</vlistitemtitle>
          </vlistitem>
        </vlistitemgroup>
      </vlist>
    </vmenu>
  </vspeeddial>
  <vwindow class=\\"v-window--main\\">
    <div data-id=\\"window-load-data\\">
      <vwindowitem value=\\"load-data\\">
        <vcontainer>
          <vcard flat=\\"\\" class=\\"pa-2\\">
            <vcardtitle class=\\"text-h5 font-weight-bold justify-center mb-3\\">

            </vcardtitle>
            <vcardtext>
              <vrow>
                <vcol align=\\"center\\">
                  <p class=\\"body-1\\">

                  </p>
                  <p class=\\"body-1\\"></p>
                  <p class=\\"disclaimer\\">

                  </p>
                </vcol>
              </vrow>
              <vrow>
                <vcol align=\\"center\\">
                  <div>
                    <!---->
                    <vrow>
                      <vcol align=\\"center\\">
                        <div>
                          <div class=\\"uppy-Root\\" dir=\\"ltr\\">
                            <div class=\\"uppy-Dashboard uppy-Dashboard--animateOpenClose\\" data-uppy-theme=\\"light\\" data-uppy-num-acquirers=\\"0\\" data-uppy-drag-drop-supported=\\"true\\" aria-hidden=\\"false\\" aria-disabled=\\"false\\" aria-label=\\"Uppy Dashboard\\">
                              <div aria-hidden=\\"true\\" class=\\"uppy-Dashboard-overlay\\" tabindex=\\"-1\\"></div>
                              <div class=\\"uppy-Dashboard-inner\\" aria-modal=\\"false\\" style=\\"width: 750px; height: 300px;\\">
                                <div class=\\"uppy-Dashboard-innerWrap\\">
                                  <div class=\\"uppy-Dashboard-dropFilesHereHint\\">Drop your files here</div>
                                  <div class=\\"uppy-Dashboard-AddFiles\\"><input class=\\"uppy-Dashboard-input\\" hidden=\\"\\" aria-hidden=\\"true\\" tabindex=\\"-1\\" type=\\"file\\" name=\\"files[]\\" multiple=\\"\\" accept=\\"\\"><input class=\\"uppy-Dashboard-input\\" hidden=\\"\\" aria-hidden=\\"true\\" tabindex=\\"-1\\" webkitdirectory=\\"true\\" type=\\"file\\" name=\\"files[]\\" multiple=\\"\\" accept=\\"\\">
                                    <div class=\\"uppy-Dashboard-AddFiles-title\\">Drop files here or <button type=\\"button\\" class=\\"uppy-u-reset uppy-Dashboard-browse\\" data-uppy-super-focusable=\\"true\\">browse files</button></div>
                                    <div class=\\"uppy-Dashboard-AddFiles-info\\"></div>
                                  </div>
                                  <div class=\\"uppy-Dashboard-progressindicators\\"></div>
                                </div>
                              </div>
                            </div>
                          </div>
                        </div>
                      </vcol>
                    </vrow>
                    <vrow>
                      <vcol class=\\"d-flex align-center justify-center\\">
                        <div class=\\"unit-files__decrypt-button\\">
                          <vdialog width=\\"500\\" scrollable=\\"\\" style=\\"z-index: 2000;\\">
                            <vcard>
                              <vcardtitle></vcardtitle>
                              <vdivider></vdivider>
                              <vcardtext>
                                <vtextfield clearable=\\"\\"></vtextfield>
                                <vtextfield clearable=\\"\\"></vtextfield>
                              </vcardtext>
                              <vdivider></vdivider>
                              <vcardactions>
                                <vspacer></vspacer>
                                <vbtn color=\\"primary\\" text=\\"\\">
                                  Close
                                </vbtn>
                              </vcardactions>
                            </vcard>
                          </vdialog>
                        </div>
                        <div class=\\"unit-files__explore-button\\">
                          <vtooltip disabled=\\"disabled\\" left=\\"true\\" class=\\"my-sm-2 mr-sm-4\\"> <span></span></vtooltip>
                        </div>
                        <div class=\\"unit-files__files-button\\">
                          <div class=\\"files-dialog\\">
                            <vdialog width=\\"500\\" scrollable=\\"\\" style=\\"z-index: 2000;\\">
                              <vcard>
                                <vcardtitle></vcardtitle>
                                <vdivider></vdivider>
                                <vcardtext>
                                  <p class=\\"mt-4\\">

                                  </p>
                                  <p class=\\"mt-4\\">

                                    <vicon>$vuetify.icons.mdiFileMultipleOutline</vicon>

                                  </p>
                                </vcardtext>
                                <vdivider></vdivider>
                                <vcardactions>
                                  <vspacer></vspacer>
                                  <vbtn color=\\"primary\\" text=\\"\\">
                                    Close
                                  </vbtn>
                                </vcardactions>
                              </vcard>
                            </vdialog>
                          </div>
                        </div>
                      </vcol>
                    </vrow>
                    <vrow>
                      <vcol>
                        <!---->
                      </vcol>
                    </vrow>
                  </div>
                </vcol>
              </vrow>
            </vcardtext>
          </vcard>
          <!---->
        </vcontainer>
      </vwindowitem>
    </div>
    <div data-id=\\"window-file-explorer\\">
      <vwindowitem value=\\"file-explorer\\" disabled=\\"disabled\\">
        <!---->
        <!---->
      </vwindowitem>
    </div>
    <div data-id=\\"window-understand-data\\">
      <!---->
    </div>
  </vwindow>
  <vbottomnavigation color=\\"secondary\\" app=\\"\\">
    <vtooltip disabled=\\"disabled\\" top=\\"true\\"> <span></span></vtooltip>
    <vtooltip disabled=\\"disabled\\" top=\\"true\\"> <span></span></vtooltip>
    <vtooltip disabled=\\"disabled\\" top=\\"true\\"> <span></span></vtooltip>
  </vbottomnavigation>
</div>"
`;
